File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-4.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        int low = Math.min(p.val, q.val);
        int high = Math.max(p.val, q.val);
        TreeNode cur = root;
        while (cur != null) {
            if (cur.val < low) {
                cur = cur.right;
            } else if (cur.val > high) {
                cur = cur.left;
            } else {
                return cur;
            }
        }
        return null;
    }
}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.max_sum = float('-inf');
        self.dfs(root)
        return self.max_sum
    
    def dfs(self, root):
        if not root:
            return 0
        
        left_sum = max(self.dfs(root.left), 0)
        right_sum = max(self.dfs(root.right), 0)
        priceNewpath = root.val + left_sum + right_sum
        self.max_sum = max(priceNewpath, self.max_sum)
        return root.val + max(left_sum, right_sum)
